include/aws/common/array_list.inl

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/* This is implicitly included, but helps with editor highlighting */
+#include <aws/common/array_list.h>
+/*
+ * Do not add system headers here; add them to array_list.h. This file is included under extern "C" guards,
+ * which might break system headers.
+ */
+
+
+AWS_STATIC_IMPL
+int aws_array_list_init_dynamic(
```

Comment:
Note that, while I put some of the resize/allocation routines out-of-line, making the initializers inline means the compiler can have visibility into the `list->item_size` element in other inlined methods (particularly given the `AWS_RESTRICT` annotation). This may allow it to make further optimizations, such as eliminating multiplications in favor of cheaper scaled-displacement addressing modes.